window.onload = () => {
  function submitOnEnter(event) {
      if (event.key !== 'Enter') {
        return;
      }
      login();
      event.preventDefault();
  }
  document.getElementById('username').addEventListener('keyup', submitOnEnter);
  document.getElementById('password').addEventListener('keyup', submitOnEnter);
}

{{> utils.js }}

async function login() {
  const username = elementGetString(document.getElementById('username'));
  const password = elementGetString(document.getElementById('password'));
  const errorMessageEntryElement = document.getElementById('errorMessageEntry');
  const errorMessageElement = document.getElementById('errorMessage');
  const loginButtonElement = document.getElementById('login');

  errorMessageEntryElement.classList.add('collapse');
  errorMessageElement.classList.remove('error');
  loginButtonElement.disabled = true;

  const progress = startProgress(loginButtonElement, 'Login');

  try {
    if (username === '') {
      throw 'Enter username';
    }
    if (password === '') {
      throw 'Enter password';
    }
    const res = await fetch('{{localUrl '/login'}}',
      {
        method: 'POST',
        headers: {
          'content-type': 'application/json'
        },
        body: JSON.stringify({
          username,
          password
        })
      }
    );
    if (res.ok || res.status === 401) {
      const json = await res.json().catch(e => { throw 'Unknown login error' });
      if (json.result === 'OK') {
        window.location = json.redirectTo ?? '{{localUrl '/'}}'
      } else {
        throw json.error ?? 'Unknown login error';
      }
    } else {
      throw `Login returned unexpected response: ${res.status} ${res.statusText}`;
    }
  } catch(e) {
    errorMessageElement.textContent = e;
    errorMessageEntryElement.classList.remove('collapse');
    errorMessageElement.classList.add('error');
  } finally {
    loginButtonElement.disabled = false;
    progress.stop('Login');
  }
}
